packages/ipfs/src/core/components/dag/resolve.js

```diff
@@ -1,16 +1,45 @@
 'use strict'
 
+const CID = require('cids')
 const { parseArgs } = require('./utils')
 const { withTimeoutOption } = require('../../utils')
 
 module.exports = ({ ipld, preload }) => {
-  return withTimeoutOption(async function * resolve (cid, path, options) { // eslint-disable-line require-await
+  return withTimeoutOption(async function resolve (cid, path, options) { // eslint-disable-line require-await
     [cid, path, options] = parseArgs(cid, path, options)
 
     if (options.preload !== false) {
       preload(cid)
     }
+    const { signal, timeout } = options
 
-    yield * ipld.resolve(cid, path, { signal: options.signal })
+    try {
+      let lastCid = cid
+      let lastRemainderPath = path
+
+      if (path) {
+        for await (const { value, remainderPath } of ipld.resolve(cid, path, {
+          signal,
+          timeout
+        })) {
+          if (!CID.isCID(value)) {
+            break
+          }
+
+          lastRemainderPath = remainderPath
+          lastCid = value
+        }
+      }
+
+      return {
+        value: lastCid,
+        remainderPath: lastRemainderPath || ''
+      }
+
+    } finally {
+      if (release) {
```

Comment:
Where is `release` defined?

Reply:
`release` was supposed to be initiated by `release = gcLock.readLock()` which i picked up from [./put.js](https://github.com/ipfs/js-ipfs/blob/master/packages/ipfs/src/core/components/dag/put.js#L45). But now I dont think it makes sense here. So removed this block.

Reply:
It's in `ipfs.dag.put` because adding blocks and pinning are async operations so gc could run between them which would result in the blocks being removed immediately after they were added.  

We're only reading here, not writing so the lock is unnecessary.